okhttp-urlconnection/build.gradle

```diff
@@ -1,9 +1,19 @@
 apply plugin: 'me.champeau.gradle.japicmp'
 
 jar {
-  manifest {
-    attributes('Automatic-Module-Name': 'okhttp3.urlconnection')
-  }
+  // modify these lines for MANIFEST.MF properties or for specific bnd instructions
+  // because this project uses the same package name as the default okhttp project (which is mostly "invalid" for OSGi)
+  // we have to remove okhttp from the bnd classpath and import it's packages manually
```

Comment:
what does OSGi want us to do in this case? Should we just say this package doesn’t exist for OSGi users? It’s pretty negligible in terms of value.

Reply:
There are 2 solutions to this:
- Rename the package name (OSGi preferred way - breaks backwards compatibility)
- Make okttp-urlconnection a fragment bundle which will at runtime basically mean that the root okhttp bundle and the okhttp-urlconnection bundle use the same ClassLoader. This means that externally when both bundles are present it's like "a single bundle". Fragment bundles are not really best practices and should generally be avoided if possible but it might work for this case.

Reply:
Can we make all of OkHttp’s libraries a fragment bundle and get the `internal` visibility we crave?

Aside from architectural purity, what’s the drawback of fragment bundles?

Reply:
Fragment bundles are kind of niche and behave differently based on the environment on some cases. PDE for example the OSGi build system from eclipse (the eclipse version of "bnd") will fail to honor packages **added** by a fragment module which is not defined on the host module. To make things work for PDE we would additionally need the "Eclipse-ExtensibleAPI: true" attribute in the manifest.

For the Okhttp use-case fragment bundles might work because every sibling module depends on the okhttp core library anyway. Do you know if any popular thirdparty library depend on one or more internal packages? (e.g. retrofit?) If that's the case we would need to export these packages anyway.

Reply:
OkHttp’s `internal` stuff is OkHttp-only.

Another option – we can stop reaching across submodules to do internal stuff? That might be simpler.

Reply:
... turns out that’s a nonstarter for MockWebServer.

Reply:
Getting rid of the internal stuff would be indeed the best solution. For okhttp-urlconnection we would still need a fragment because of the package name issue.

Btw. MockWebServer - I have missed that module, I have to modify the build.gradle for that one.

Maybe a hybrid solution would be possible? For those modules still requiring internal apis we use the fragment approach and for the other modules we create a "normal" OSGi bundle.